### PR TITLE
feat: chainlink oracle adapter

### LINKED ISSUE #993: Add HDX chainlink oracle adapter
We need to add oracle for money market to accept HDX

### PR DESCRIPTION
**Pull request by Roznovjak**
Fixes #993

### CODE REVIEW

File: runtime/hydradx/src/evm/mod.rs [hunk 1]
```diff
@@ -121,6 +121,8 @@ impl Get<Multiplier> for TransactionPaymentMultiplier {
 	}
 }
 
+pub const USDT: AssetId = 10u32;
```
[the remark this answers is not in the document]
Comment (Roznovjak) [reply runtime/hydradx/src/evm/mod.rs:124]: Imho this is not necessary. We don't expect asset id's to change. Using asset registry also adds storage read.

File: runtime/hydradx/src/evm/precompiles/chainlink_adapter.rs [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+		})?
+		.checked_div(price.d.into())
+		.ok_or(PrecompileFailure::Error {
+			exit_status: pallet_evm::ExitError::Other("Price conversion failed.".into()),
+		})
+}
+
+/// Encoding is 7 bytes for precompile prefix 0x00000000000001,
+/// followed by 1 byte for encoded OraclePeriod enum, 8 bytes for Source, and 4 bytes for AssetId.
+pub fn encode_evm_address(asset_id: AssetId, period: OraclePeriod, source: Source) -> Option<EvmAddress> {
```
[the remark this answers is not in the document]
Comment (Roznovjak) [reply runtime/hydradx/src/evm/precompiles/chainlink_adapter.rs:227]: Do you propose to write prop tests or simpler tests with different edge cases? One thing I don't like is that in prop tests you need to manually list Enums. So if someone adds new item to the enum, tests are not aware of it.

File: runtime/hydradx/src/evm/precompiles/chainlink_adapter.rs [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+			};
+		}
+		Err(PrecompileFailure::Revert {
+			exit_status: ExitRevert::Reverted,
+			output: "invalid price oracle data".into(),
+		})
+	}
+}
+
+impl<QuoteAsset, Runtime> ChainlinkOraclePrecompile<QuoteAsset, Runtime>
```
[the remark this answers is not in the document]
Comment (Roznovjak) [reply runtime/hydradx/src/evm/precompiles/chainlink_adapter.rs:89]: It's the interface used by DIA oracle contract.

File: runtime/hydradx/src/evm/precompiles/chainlink_adapter.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (mrq1911) [reply runtime/hydradx/src/evm/precompiles/chainlink_adapter.rs:227]: `encode_precompile_address` or `encode_precompile_address`, any address is evm address, evm is redundant term

[the remark this answers is not in the document]
Comment (mrq1911) [reply runtime/hydradx/src/evm/precompiles/chainlink_adapter.rs:227]: but based on `is_oracle_address` this should be `encode_oracle_address`

[the remark this answers is not in the document]
Comment (mrq1911) [reply runtime/hydradx/src/evm/precompiles/chainlink_adapter.rs:227]: also it would be preferable for this to not return Option as there is not really reason to do so

[the remark this answers is not in the document]
Comment (Roznovjak) [reply runtime/hydradx/src/evm/precompiles/chainlink_adapter.rs:227]: ok, I will remove Option and write a test to ensure that encoded size of OraclePeriod is 1 byte.

**Review by mrq1911 — APPROVED** [review]

File: runtime/hydradx/src/evm/mod.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (mrq1911) [reply runtime/hydradx/src/evm/mod.rs:124]: due to later changes this is irrelevant
